pyls/rpc_manager.py

```diff
@@ -0,0 +1,171 @@
+# Copyright 2017 Palantir Technologies, Inc.
+import logging
+from uuid import uuid1
+
+from concurrent.futures import ThreadPoolExecutor, Future
+from jsonrpc.base import JSONRPCBaseResponse
+from jsonrpc.jsonrpc1 import JSONRPC10Response
+from jsonrpc.jsonrpc2 import JSONRPC20Response, JSONRPC20Request
+from jsonrpc.exceptions import JSONRPCMethodNotFound, JSONRPCInternalError
+
+log = logging.getLogger(__name__)
+
+RESPONSE_CLASS_MAP = {
+    "1.0": JSONRPC10Response,
+    "2.0": JSONRPC20Response
+}
+
+
+class JSONRPCManager(object):
+    """ Implementation of the Microsoft VSCode Language Server Protocol
+    https://github.com/Microsoft/language-server-protocol/blob/master/versions/protocol-1-x.md
+    """
+
+    def __init__(self, message_manager, message_handler):
+        self._message_manager = message_manager
+        self._message_handler = message_handler
+        self._shutdown = False
+        self._sent_requests = {}
+        self._received_requests = {}
+        self._executor_service = ThreadPoolExecutor()
+
+    def start(self):
+        """Start reading JSONRPC messages off of rx"""
+        self.consume_requests()
+
+    def shutdown(self):
+        """Set flag to ignore all non exit messages"""
+        self._shutdown = True
+
+    def exit(self):
+        """Stop listening for new message"""
+        self._executor_service.shutdown()
+        self._message_manager.close()
+
+    def call(self, method, params=None):
+        """Send a JSONRPC message with an expected response.
+
+        Args:
+            method (str): The method name of the message to send
+            params (dict): The payload of the message
+
+        Returns:
+            Future that will resolve once a response has been recieved
+
+        """
+        log.debug('Calling %s %s', method, params)
+        request = JSONRPC20Request(_id=uuid1().int, method=method, params=params)
+        request_future = Future()
+        self._sent_requests[request._id] = request_future
+        self._message_manager.write_message(request.data)
+        return request_future
+
+    def notify(self, method, params=None):
+        """Send a JSONRPC notification.
+
+         Args:
+             method (str): The method name of the notification to send
+             params (dict): The payload of the notification
+         """
+        log.debug('Notify %s %s', method, params)
+        notification = JSONRPC20Request(method=method, params=params)
+        self._message_manager.write_message(notification.data)
+
+    def cancel(self, request_id):
```

Comment:
You don't want/need this. Instead, you should add a done callback in the `call` method that checks if the future is cancelled, and if so sends the cancel message. That way a user can call `manager.call('method').cancel()`

Reply:
Thats actually not true. This doesn't let you operate on requests send from the language server to the client(whether we should is a totally separate question). The reason for having this method is to allow the language server correctly act upon `$/cancel` requests send from the client